lldb/tools/lldb-dap/Handler/AttachRequestHandler.cpp

```diff
@@ -29,38 +29,43 @@ namespace lldb_dap {
 ///
 /// Since attaching is debugger/runtime specific, the arguments for this request
 /// are not part of this specification.
-Error AttachRequestHandler::Run(const AttachRequestArguments &args) const {
+void AttachRequestHandler::Run(const AttachRequestArguments &args,
+                               Callback<Error> callback) const {
   // Initialize DAP debugger and related components if not sharing previously
   // launched debugger.
   std::optional<int> debugger_id = args.debuggerId;
   std::optional<lldb::user_id_t> target_id = args.targetId;
 
   // Validate that both debugger_id and target_id are provided together.
-  if (debugger_id.has_value() != target_id.has_value()) {
-    return llvm::createStringError(
+  if (debugger_id.has_value() != target_id.has_value())
+    return callback(llvm::createStringError(
         "Both debuggerId and targetId must be specified together for debugger "
-        "reuse, or both must be omitted to create a new debugger");
-  }
+        "reuse, or both must be omitted to create a new debugger"));
```

Comment:
For a `DAPError`, we delay the response until the configuration is done. But these we still return immediately. The distinction seems arbitrary. How am I as a developer to know whether my error should be a DAPError or not? Should we delay all errors and if so, can we make the API easier to use correctly and harder to use incorrectly (i.e. by abstracting the callback away behind a DelayResponse abstraction or something)?

Reply:
I removed the generic AsyncRequestHandler with a callback and instead made a specific base class for attach/launch that handles these two requests and delays their response until configurationDone. This makes it easier to follow and reverts most of the changes to the `AttachRequestHandler::Run` and `LaunchRequestHandler::Run`.